### PR TITLE
utils: allow developers to differ in ruby versions

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This should allow some testing with ruby 2.7 before macOS actually switches.

### CODE REVIEW

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby_check_version_script.rb:20]: I wonder if we should allow major version deviation too; Ruby 3.0 isn't actually that incompatible with Ruby 2.6 (compared to Ruby 2.5).

Relatedly, may be worth allowing `HOMEBREW_DEVELOPER` to use higher versions but not lower ones. This is best done with a `Gem::Version >= Gem::Version` comparison.

**Review by MikeMcQuaid — APPROVED** [review]
Fine as-is but some comments I'd like to see addressed in a quickish follow-up after this is merged (or on this PR).

**Comment by Bo98** [comment]
> This should allow some testing with ruby 2.7 before macOS actually switches.

We'll maybe find out in a couple weeks but I honestly doubt macOS will switch to 2.7 - it's EOL in less than a year's time.

If macOS remains on 2.6 and we decide to leave it behind, we'll probably pick 3.1 for our Portable Ruby.

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (Bo98) [reply Library/Homebrew/utils/ruby_check_version_script.rb:20]: I would be surprised if the kwarg changes in Ruby 3.0 didn't cause any issues. Some of those issues may also be subtle.

**Comment by SMillerDev** [comment]
Yeah, I'm fine with Mike's suggestion to allow anything higher. Just wanted to start the PR with a small step.

**Comment by MikeMcQuaid** [comment]
> If macOS remains on 2.6 and we decide to leave it behind, we'll probably pick 3.1 for our Portable Ruby.

I would rather we kept on the system Ruby for our macOS users even if that means sticking on 2.6.

The only >=2.7 need I see right now is for Tapioca which is a particularly niche maintainers case.

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/ruby_check_version_script.rb:20]: @Bo98 🤷🏻 I feel like that's why this is an undocumented, developer-only thing.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/ruby_check_version_script.rb:20]: Although this does make me think it's probably better to adding a undocumented `HOMEBREW_ALLOW_ANY_RUBY_VERSION` that is required as well as `HOMEBREW_DEVELOPER` to avoid screwing with people and CI where `HOMEBREW_DEVELOPER` is set.

**Comment by Bo98** [comment]
> The only >=2.7 need I see right now is for Tapioca which is a particularly niche maintainers case.

Tapioca runs external to `brew` so no Ruby version checks need changing for that. A second installation of bundler is all that's needed.

**Comment by MikeMcQuaid** [comment]
> > I would rather we kept on the system Ruby for our macOS users even if that means sticking on 2.6.
> 
> I should have also said "or removes system Ruby" too. We'll need to decide if moving to ship with Xcode rather than the system (like Python) qualifies as that (we don't currently require dev tools to be installed for bottle-only installations).

Yes. If it's just as available as currently: we'll avoid shipping a portable Ruby. It may also be we avoid shipping if it a CLT/Xcode is installed.

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (carlocab) [inline Library/Homebrew/utils/ruby_check_version_script.rb:20]:
```suggestion
if (ENV["HOMEBREW_DEVELOPER"].present? && homebrew_required_ruby_version > ruby_version) ||
```
Per the docs you should be able to set this to anything other than the empty string.

Unless we want to gate this specifically to users who have this set to `1`, that is.

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (SMillerDev) [reply Library/Homebrew/utils/ruby_check_version_script.rb]: So... Yay or nay on the current version?

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/utils/ruby_check_version_script.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby_check_version_script.rb:20]:
```suggestion
if (ENV["HOMEBREW_DEVELOPER"].present? && ENV["HOMEBREW_USE_RUBY_FROM_PATH"].present? && homebrew_required_ruby_version > ruby_version) ||
```

**Review by MikeMcQuaid — COMMENTED** [review]
One change and then 👍🏻.

As a thought, for consistency I wonder if `HOMEBREW_FORCE_BREWED_RUBY` might be nice which just uses whatever the `ruby` formula version is? Any thoughts anyone?

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by Bo98** [comment]
`HOMEBREW_USE_RUBY_FROM_PATH` isn't really what's happening here though. This is just a version check and macOS is still hardcoded to only use `/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby` (and I think even Linux can have issues finding things outside `/usr/bin`, but maybe that was fixed recently).

**Comment by SMillerDev** [comment]
On Linux this definitely works from path.

**Comment by Bo98** [comment]
Ok good to know that works now. Definitely had some user reports about that not being the case at some point.

**Comment by Bo98** [comment]
> We should make sure you don't end up breaking things with a `brew upgrade ruby` though.

To support a rolling Ruby target given we'd need to vendor gems for all Ruby versions, unless we force `brew install-bundler-gems`, or rather a subset which doesn't invoke any native compilation ideally.

**Comment by MikeMcQuaid** [comment]
> To support a rolling Ruby target given we'd need to vendor gems for all Ruby versions, unless we force `brew install-bundler-gems`, or rather a subset which doesn't invoke any native compilation ideally.

For developers using an undocumented flag: I don't think that's necessary. We don't want to support this for end-users.

**Comment by Bo98** [comment]
> > To support a rolling Ruby target given we'd need to vendor gems for all Ruby versions, unless we force `brew install-bundler-gems`, or rather a subset which doesn't invoke any native compilation ideally.
> 
> For developers using an undocumented flag: I don't think that's necessary. We don't want to support this for end-users.

To me what makes sense then is a flag solely focussed on overriding the version check rather than attaching it to things like `HOMEBREW_RUBY_PATH` or `HOMEBREW_FORCE_BREWED_RUBY`.

Using a custom Ruby install is to me a valid end-user case for platforms we don't ship portable Ruby, but allowing unsupported Ruby versions should never be.

**Comment by MikeMcQuaid** [comment]
> Using a custom Ruby install is to me a valid end-user case for platforms we don't ship portable Ruby

This already works, we use a working Ruby from the PATH?

**Comment by Bo98** [comment]
> > Using a custom Ruby install is to me a valid end-user case for platforms we don't ship portable Ruby
> 
> This already works, we use a working Ruby from the PATH?

My point is though: `HOMEBREW_USE_RUBY_FROM_PATH` is confusing to me since it sounds like it's not like that by default.

**Comment by MikeMcQuaid** [comment]
I'm fine with either.

**Comment by SMillerDev** [comment]
@Bo98 @MikeMcQuaid are you fine with me merging like this and making a macOS PR later?

**Comment by MikeMcQuaid** [comment]
@SMillerDev Hold for now. Thanks for the PR but want to check a few things first.

Can you explain what you see this PR achieving and on what platforms? It mentions macOS and 2.7 but we don't have a new Ruby version on Ventura yet and it's scoped to Linux.

**Comment by SMillerDev** [comment]
> Can you explain what you see this PR achieving and on what platforms? It mentions macOS and 2.7 but we don't have a new Ruby version on Ventura yet and it's scoped to Linux.

Obviously on macOS we have to deal with the system ruby, but that version has been EOL for months already. Initially my goal for this was to allow testing with the next minor version to allow developers to spot brew bugs before the general public does, but it has since expanded to testing "any newer version". I initially tested this on Linux since my desktop system had ruby 2.7 already but later learned that macOS needs some more work to allow this to work there too.

Since this is hidden behind an undocumented environment variable I don't expect this to break any existing code, so I would say the macOS change can be in a new PR.

**Comment by MikeMcQuaid** [comment]
> but it has since expanded to testing "any newer version".

Does it just use any newer version in the `PATH`?

Does it handle e.g. `brew install-bundler-gems` as expected?

> Since this is hidden behind an undocumented environment variable I don't expect this to break any existing code, so I would say the macOS change can be in a new PR.

I think ideally they'd be in the same PR but if you're planning on opening a macOS PR right afterwards: fine with me.

**Comment by SMillerDev** [comment]
> Does it just use any newer version in the PATH?

Yeah, or more accurately the current change allows the ruby version to be newer (it's only missing the change to allow ruby from path on macOS).

> Does it handle e.g. brew install-bundler-gems as expected?

It's quite a mess with 2 ruby versions, but it works correctly, yes.

**Comment by SMillerDev** [comment]
I knew I shouldn't have complained about that.

IIRC my main complaint was the way ruby installs all gems in a version dependent folder, so if you try this with a couple versions you have 4 installs of the ruby universe in the vendor directory.

**Comment by Bo98** [comment]
I'm a bit lost between conversations - yesterday I thought we weren't supporting newer Rubies. Looks like we now are again.

> IIRC my main complaint was the way ruby installs all gems in a version dependent folder, so if you try this with a couple versions you have 4 installs of the ruby universe in the vendor directory.

One install per Ruby minor version is expected. We could symlink those we've vendored to allow Homebrew to actually work on these newer versions without having to work with bundler outside of `brew`, but anything with any sort of native extension installed by `install-bundler-gems` should not be symlinked and have separate installs.

Ruby 3.x will likely require code changes due to kwargs changes. Ruby 2.7 is a weird middle that has quirks that are sometimes incompatible with 2.6, sometimes incompatible with 3.x so need to be cautious about that one.

File: Library/Homebrew/utils/ruby.sh
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby.sh:20]:
```suggestion
  if [[ -n "${HOMEBREW_MACOS}" && -z "${HOMEBREW_USE_RUBY_FROM_PATH}" ]]
```

Presumably?

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**
